### PR TITLE
[Bug] Fix bug that prevented custom tables/grids with column referral

### PR DESCRIPTION
**Pull request by maxschulz-COL**
## Description
Closes https://github.com/mckinsey/vizro/issues/435 and https://github.com/McK-Internal/vizro-internal/issues/747

We now send the full DF during the build phase (instead of an empty `pd.DataFrame`). This could have performance consequences for very large DFs, but on the grand scheme of things I think it probably the better solution.

And alternative would be to send an empty DF with the columns present, but that may lead to other bugs down the road.

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

**Comment by huong-li-nguyen** [comment]
Thanks for fixing so quickly! 🚀  What "potential bugs down the road" do you imagine? I actually like the alternative solution approach - it's also the one we've implemented on VizX actually. I think it did not lead to any major bugs, and the performance issue was more severe.

**Comment by maxschulz-COL** [comment]
> Thanks for fixing so quickly! 🚀 What "potential bugs down the road" do you imagine? I actually like the alternative solution approach - it's also the one we've implemented on VizX actually. I think it did not lead to any major bugs (as far as I remember), and the performance issue was more severe.

So of the top of my head I could imagine that someone not only refers to a column, but also a specific row. This then would fail, as there are no rows present.

This seems to be a case of a wider range of things where people refer or rely on the size of the data, or generally anything regarding the original DF they provide.

One could argue that it is bad design to write a custom grid/table like that, but the confusion of the user showed that people simply do not expect the data to be different at any point from what they originally provide

> Do we know how other tools deal with this? e.g. do they load in the data in batches? or do they always load in the entire dataset?

So the AgGrid for example has the opportunity to [enable batch loading in infinite scroll](https://www.ag-grid.com/javascript-data-grid/infinite-scrolling/#turning-on-infinite-scrolling), but I think that defeats a little bit the point. In principle we are not converned with loading the entire data, that is what we do anyway once the page loads with all filters/parameters etc. It is simply this initial building that immediately gets overwritten.

Ideally we would want to just create an entirely different loading component (like for `graph`), but it has been shown (see comment above the build lines in both table and grid) that having different settings for the objects causes problems (the case we observed was the pagination setting). Remember the long discussion we had where you said, just write "Do not change this line" :). So this is related to that.

**Comment by huong-li-nguyen** [comment]
Thanks for clarifying!

File: vizro-core/src/vizro/models/_components/ag_grid.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def build(self):
                 # here) must have the same setting as the object that is built by the on-page-load mechanism using
                 # with the user settings and rendered finally. Otherwise the grid is not rendered correctly.
                 # Hence be careful when editing the line below.
-                html.Div(self.__call__(data_frame=pd.DataFrame()), id=self.id, className="table-container"),
+                html.Div(self.__call__(), id=self.id, className="table-container"),
```
Comment (petar-qb) [inline vizro-core/src/vizro/models/_components/ag_grid.py:114]: Maybe something like this could work:

```
 def build(self):
        clientside_callback(...)
        grid = self.__call__()
        grid["rowData"] = {}  # or `grid["rowData"] = pd.DataFrame()`
        return dcc.Loading([..., html.Div(grid, ...), ...], ...)
```

This should send the original data_frame into the `figure` callable, but also return an empty DataFrame with all calculated configuration on the `build load`. Then the UI chart goes into the Loading state , and finally the `update_figures` returns filtered data + parametrised configuration.

Comment (maxschulz-COL) [reply vizro-core/src/vizro/models/_components/ag_grid.py:114]: It is an interesting idea that I will try out

**Comment by antonymilne** [comment]
I would like to review this so please don't merge yet 🙂

File: vizro-core/src/vizro/models/_components/table.py [hunk 2]
```diff
@@ -108,7 +108,7 @@ def build(self):
             html.Div(
                 [
                     html.H3(self.title, className="table-title") if self.title else None,
-                    html.Div(self.__call__(data_frame=pd.DataFrame()), id=self.id),
+                    html.Div(self.__call__(), id=self.id),
```
Comment (antonymilne) [inline vizro-core/src/vizro/models/_components/table.py:111]: Please could you add an explanatory comment here that refers to the one in `AgGrid.build` or just copies and pastes it in? (I don't know if the pagination stuff is also relevant here or if it's just done the same way to match AgGrid, but I was confused just reading through the code why we did this here rather than just returning an empty `DataTable`)

**Review by antonymilne — APPROVED** [review]
tl;dr: thanks @petar-qb and @maxschulz-COL for your great work on this. Hopefully longer term we will somehow have an entirely better system here but for now this looks good.

---

I prefer the solution @petar-qb suggested in https://github.com/mckinsey/vizro/pull/439/files#r1580918187 so would be great if you could try it out. But the current solution is fine too.

I am not hugely concerned right now about the performance hit this incurs, but I am not keen on it either. In general I don't much like the "double loading" we currently have where we do everything once with an empty dataframe (or the real thing, like now) and then immediately override it with the real data. This PR unfortunately, but necessarily, adds another layer of unsatisfactoriness to that scheme 😬 

That isn't meant as a criticism of this PR or the existing system, since I know the current on-page-load system has its merits and lots of good reasoning behind it. I tried to come up with some improvements before and couldn't. I just hope that as part of the actions v2 work we can somehow improve this scheme though 🤞

File: vizro-core/src/vizro/models/_components/ag_grid.py (hunk 1 above)
Comment (maxschulz-COL) [reply vizro-core/src/vizro/models/_components/ag_grid.py:116]: I tried this in https://github.com/mckinsey/vizro/pull/439/commits/4c3924653c2db277d9200dbd010005d043a1a6f6, and it _almost works_.

There seem to be no issues with pagination for example, but unfortunately it seems to break the persistence settings.

Presumably that is because it "forgets" the filters when an intermediate AGGrid has no data. You can check out the complete changes at the commit above. Try setting a filter for the first AGGrid and change page. Upon returning the filter is gone, although persistence is set to true.

Comment (antonymilne) [reply vizro-core/src/vizro/models/_components/ag_grid.py:116]: Ah ok, thanks for trying it out! I'm fine with your current solution then. Hopefully we can improve this some time in the future.

**Review by antonymilne — APPROVED** [review]
